cmd/influxd/inspect/tsm_verify.go

```diff
@@ -85,65 +27,80 @@ type verifyTSM struct {
 	err   error
 }
 
-func (v *verifyTSM) loadFiles(dataPath string) error {
-	err := filepath.Walk(dataPath, func(path string, f os.FileInfo, err error) error {
-		if err != nil {
-			return err
-		}
-		if filepath.Ext(path) == "."+tsm1.TSMFileExtension {
-			v.files = append(v.files, path)
-		}
-		return nil
-	})
-
-	if err != nil {
-		return errors.Wrap(err, "could not load storage files (use -dir for custom storage root)")
-	}
+type verifyUTF8 struct {
+	verifyTSM
+	totalErrors int
+	total       int
+}
 
-	return nil
+type verifyChecksums struct {
+	verifyTSM
+	totalErrors int
+	total       int
 }
 
-func (v *verifyTSM) Next() bool {
-	if len(v.files) == 0 {
-		return false
+func NewTSMVerifyCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   `verify-tsm`,
+		Short: `Verifies the integrity of TSM files`,
+		Args:  cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var runner verifier
+			if checkUTF8 {
+				runner = &verifyUTF8{}
+			} else {
+				runner = &verifyChecksums{}
+			}
+			err := runner.Run(cmd, dir)
+			return err
+		},
 	}
-
-	v.f, v.files = v.files[0], v.files[1:]
-	return true
+	cmd.Flags().StringVar(&dir, "dir", os.Getenv("HOME")+"/.influxdbv2", "Root storage path.")
+	cmd.Flags().BoolVar(&checkUTF8, "check-utf8", false, "Verify series keys are valid UTF-8. This check skips verification of block checksums.")
+	cmd.Flags().BoolVar(&verbose, "verbose", false, "Enable verbose logging")
+	return cmd
 }
 
-func (v *verifyTSM) TSMReader() (string, *tsm1.TSMReader) {
-	file, err := os.OpenFile(v.f, os.O_RDONLY, 0600)
-	if err != nil {
-		v.err = err
-		return "", nil
+func (v *verifyUTF8) Run(cmd *cobra.Command, dataPath string) error {
+	if err := v.loadFiles(dataPath); err != nil {
+		return err
 	}
 
-	reader, err := tsm1.NewTSMReader(file)
-	if err != nil {
-		file.Close()
-		v.err = err
-		return "", nil
-	}
+	v.Start()
 
-	return v.f, reader
-}
+	for v.Next() {
+		f, reader := v.TSMReader()
+		if reader == nil {
+			break
+		}
 
-func (v *verifyTSM) Start() {
-	v.start = time.Now()
-}
+		n := reader.KeyCount()
+		fileErrors := 0
+		v.total += n
+		for i := 0; i < n; i++ {
+			key, _ := reader.KeyAt(i)
```

Comment:
What's the 2nd return value here (if it's an error, should we do something with it)?

Reply:
It returns the key type, which for verifying valid UTF-8 chars, I don't think we need to do anything with